info-articles/Mapping in solidity - issue #19.md/article.md

```diff
@@ -0,0 +1,437 @@
+ 
+      if(!successful[_key])  {
+        successful[_key] = true;
+        keys.push(_key);
+      }
+     }
+```
+We passed the key and value pairs variables into the parameter and assigned the key of the balances to the value. It is time to set the second mapping, which has a boolean value.
+
+Recall that the default value of bool is false. Therefore, we declared the negation of the successful mapping against the key. Then we stated that the compiler should push the key into the array if a successful key is true.
+```
+  function getLength() public view returns(uint){
+        return keys.length;
+      }
+
+This function attempts to get the length of the array.
+
+  function getTheI(uint i) public view returns(uint) {
+        return theBalanceOf[keys[i]];
+      }
+```
+We will need the `i variable` when we want to run a loop. Thus, this function returned the balances of keys that form the i variable. Once you have this, you can loop through the mappings we examined earlier in this guide.
+```
+    function getGenesis() public view returns(uint) {
+        return theBalanceOf[keys[0]];
+      }
+ 
+      function getTheLast() public view returns(uint) {
+       return theBalanceOf[keys[keys.length -1]];
+      }
+```
+These two getter functions retrieved the balances of key[0], which is the array's first element, and the array's last member, which is the length minus 1.
+
+## Creating a Custom Contract in Solidity to Explain Mapping
+Since we have learned much about mapping, let us create a simple contract and use the knowledge in real time.
+```
+// SPDX-License-Identifier: MIT
+ pragma solidity ^0.8.16;
+ 
+  contract TheWallet{
+      address public John;
+ 
+      constructor() {
+          John == msg.sender;
+      }
+   
+    mapping(address => uint) public hisBalance;
+    mapping (address=> mapping(address => uint)) public allowThem;
+   
+    function deposit() public payable {
+        hisBalance[msg.sender] += msg.value;
+    }
+ 
+    function withdrawal() public payable {
+        (bool sent, ) = msg.sender.call{value: hisBalance[msg.sender]}("");
+        require(sent, "cannot withdraw");
+    }
+ 
+    function transfer(address receiver, uint token) public returns(bool){
+        allowThem[msg.sender][receiver] += token;
+        hisBalance[msg.sender] += token;
+        hisBalance[receiver] -= token;
+        return true;
+    }
+ 
+    function remainingMoney() public view returns (uint) {
+        return address(this).balance;
+    }
+  }
+ ```
+ 
+
+This contract is a simple wallet contract where users can deposit, transfer, and know the remaining Ethers in the contract. Here, we used mapping to track both balances and allowance.
+
+The next stage is to compile and deploy this contract using Hardhat. Go to the Hardhat folder on your device, locate the src folder and write your contract there. Perhaps you have not installed Hardhat in your terminal, paste this on your Vs Code terminal:

```

Comment:
I don't think there is an src folder in hardhat anymore. The smart contract needs to go under the 'contracts' folder.

Reply:
Yes, yes, discovered that too.